### PR TITLE
Add standard about phpstan annotations

### PR DESCRIPTION
**Pull request by VincentLanglet**

### CODE REVIEW

**Review by OskarStark — APPROVED** [review]

**Review by core23 — APPROVED** [review]

File: templates/project/CONTRIBUTING.md [hunk 1]
```diff
@@ -93,13 +93,18 @@ php-cs-fixer fix --verbose
 Please note that we try to keep phpdoc to a minimum, so if an `@param` phpdoc
 comment brings nothing more than the type hint and variable name already do,
 it SHOULD be removed. Descriptions are OPTIONAL if you want to document a type.
+If you want to use pseudo-types for static-analysis, prefer the usage of
```
Comment (jordisala1991) [inline templates/project/CONTRIBUTING.md:96]: pseudo types or generics (with an exception of the @template annotation

Comment (core23) [reply templates/project/CONTRIBUTING.md:96]: Why is the `@template` annotation an exception, you could/should also prefix this. AFAIK there was a symfony templating conflict with this annotation.

Comment (VincentLanglet) [reply templates/project/CONTRIBUTING.md:96]: I agree, we should always recommend tu use the phpstan prefix.

File: templates/project/CONTRIBUTING.md [hunk 2]
```diff
@@ -93,13 +93,18 @@ php-cs-fixer fix --verbose
 Please note that we try to keep phpdoc to a minimum, so if an `@param` phpdoc
 comment brings nothing more than the type hint and variable name already do,
 it SHOULD be removed. Descriptions are OPTIONAL if you want to document a type.
+If you want to use pseudo-types, generics or templates for static-analysis,
+prefer the usage of `@phpstan-` annotations over classic annotations to keep
+IDE support, and over `@psalm-` annotations for consistency.
```
Comment (core23) [inline templates/project/CONTRIBUTING.md:98]: We should keep a note, that you should keep the classic type hints and add additional type hints / generics via phpstan or psalm.

**Review by core23 — CHANGES_REQUESTED** [review]

File: templates/project/CONTRIBUTING.md (hunk 2 above)
Comment (VincentLanglet) [reply templates/project/CONTRIBUTING.md:98]: Do you have any proposal about the text of the note ? I already don't really like the big sentence I made.

File: templates/project/CONTRIBUTING.md [hunk 3]
```diff
@@ -93,13 +93,18 @@ php-cs-fixer fix --verbose
 Please note that we try to keep phpdoc to a minimum, so if an `@param` phpdoc
 comment brings nothing more than the type hint and variable name already do,
 it SHOULD be removed. Descriptions are OPTIONAL if you want to document a type.
+If you want to use pseudo-types, generics or templates for static-analysis,
```
Comment (jordisala1991) [inline templates/project/CONTRIBUTING.md:96]: please, add array shape too

**Comment by VincentLanglet** [comment]
We should start merging something @sonata-project/contributors.

We will still be able to improve the standard in the futur if needed.

**Review by jordisala1991 — APPROVED** [review]

File: templates/project/CONTRIBUTING.md.twig [hunk 4]
```diff
@@ -93,13 +93,18 @@ php-cs-fixer fix --verbose
 Please note that we try to keep phpdoc to a minimum, so if an `@param` phpdoc
 comment brings nothing more than the type hint and variable name already do,
 it SHOULD be removed. Descriptions are OPTIONAL if you want to document a type.
+If you want to use pseudo-types, generics, array shapes or templates for
+static-analysis, prefer the usage of `@phpstan-` annotations over classic
+annotations to keep IDE support, and over `@psalm-` annotations for consistency.
```
Comment (core23) [inline templates/project/CONTRIBUTING.md.twig:99]:
```suggestion
+Keep standard annotations if necessary.

```

**Review by OskarStark — DISMISSED** [review]

**Comment by OskarStark** [comment]
I agree 🤘🏻

File: templates/project/CONTRIBUTING.md.twig (hunk 4 above)
Comment (VincentLanglet) [reply templates/project/CONTRIBUTING.md.twig:101]: Done

**Review by OskarStark — APPROVED** [review]

**Review by core23 — APPROVED** [review]

**Comment by VincentLanglet** [comment]
PHPStorm does support `@psalm-type`, `@psam-import-type` and `@psalm-template` but not the `@phpstan-` annotation (and report missing classes since 2021.2) so I wonder if we should
1) Pray for a support of `@phpstan-` annotation soon
2) Change to `@psalm-` annotation

@sonata-project/contributors

**Pull request merged**